test/integration/puppetlabs/services/jruby/jruby_pool_int_test.clj

```diff
@@ -236,90 +237,149 @@
+         ;; Wait for some arbitrary (but short) period of time to see that
+         ;; that the agent is still busy - presumably handling the flush.
+         (is (false? (await-for 50 pool-agent)))
+
+         ;; and finally, we return the last instance from the old pool
+         (jruby-protocol/return-instance jruby-service instance1 :max-requests-flush-while-pool-flush-in-progress-test)
+
+         ;; wait until the flush is complete
+         (is (true? (timed-await pool-agent))
+             (str "timed out waiting for the flush to complete, stack:\n"
+                  (get-all-stack-traces-as-str))))
+
+       ;; we should have three instances with the constant and one without.
+       (is (true? (check-jrubies-for-constant-counts pool-context 3 1)))
+
+       ;; The jruby return instance calls done within the previous
+       ;; check jrubies call may cause an instance to be in the process of
+       ;; being flushed when the server is shut down.  This ensures that
+       ;; the flushing is all done before the server is shut down - since
+       ;; that could otherwise cause an annoying error message about the
+       ;; pool not being full at shut down to be displayed.
+       (timed-await (:flush-instance-agent pool-context))))))
+
+(deftest initialization-and-cleanup-hooks-test
+  (testing "custom initialization and cleanup callbacks get called appropriately"
+    (let [lifecycle-fns {:initialize (fn [instance] (assoc instance :foo "FOO"))
+                         :shutdown (fn [instance] (log/error "Terminating" (:foo instance)))}
+          config (assoc-in (jruby-testutils/jruby-tk-config
+                            (jruby-testutils/jruby-config
+                             {:max-active-instances 1
+                              :max-requests-per-instance 10
+                              :borrow-timeout default-borrow-timeout}))
+                           [:jruby :lifecycle] lifecycle-fns)]
+      (logutils/with-test-logging
+       (tk-testutils/with-app-with-config
+        app
+        [jruby/jruby-pooled-service]
+        config
+        (let [jruby-service (tk-app/get-service app :JRubyService)
+              context (tk-services/service-context jruby-service)
+              pool-context (:pool-context context)]
+          ;; set a ruby constant in each instance so that we can recognize them
+          (is (true? (set-constants-and-verify pool-context 1)))
+          (let [instance (jruby-protocol/borrow-instance jruby-service
+                                                         :initialization-and-cleanup-hooks-test)]
+            (is (= "FOO" (:foo instance)))
+            (jruby-protocol/return-instance jruby-service instance :initialization-and-cleanup-hooks-test))
+
+          (jruby-protocol/flush-jruby-pool! jruby-service)
+          ; wait until the flush is complete
+          (await (get-in context [:pool-context :pool-agent]))
+          (is (logged? #"Terminating FOO"))))))))
```

Comment:
I think we just basically copied this out of the master-termination-test https://github.com/puppetlabs/jruby-utils/pull/7/files#diff-5420d9f56ec9b26871e6cd8bdce73df9L132, but agreed that using an atom would be less fragile.